server/src/test/java/org/candlepin/model/ProductTest.java

```diff
@@ -245,4 +256,27 @@ public void testClone(String valueName, Object value1, Object value2) throws Exc
         assertEquals(base.hashCode(), clone.hashCode());
     }
 
+    @Test
+    public void testProductEqualityWithCycles() {
+        Product product = new Product();
+        Product derivedProduct = new Product();
+
+        product.setUuid("test-UUID-1");
+        derivedProduct.setUuid("test-UUID-2");
+
+        // Creating endless cyclic product hierarchy
+        product.setDerivedProduct(derivedProduct);
+        derivedProduct.setDerivedProduct(product);
+
+        assertFalse(product.equals(derivedProduct));
```

Comment:
You can use `assertNotEquals` for a bit more readable assertion on equality.

Reply:
fixed.